wandb/apis/importers/internal.py

```diff
@@ -0,0 +1,437 @@
+            progress.subtask_pbar.remove_task(task)
+
+
+def _handle_log_specific_artifact(
+    sm: sender.SendManager,
+    rm: RecordMaker,
+    art: Artifact,
+    config: SendManagerConfig,
+):
+    if config.log_artifacts:
+        sm.send(rm._make_artifact_record(art))
+
+
+def _handle_use_specific_artifact(
+    sm: sender.SendManager,
+    rm: RecordMaker,
+    art: Artifact,
+    config: SendManagerConfig,
+):
+    if config.use_artifacts:
+        sm.send(rm._make_artifact_record(art, use_artifact=True))
+
+
+def _handle_history(
+    sm: sender.SendManager,
+    rm: RecordMaker,
+    config: SendManagerConfig,
+    run_identifier: str,
+):
+    if config.history:
+        task_name = f"History {run_identifier}"
+        task = progress.subtask_pbar.add_task(task_name, total=None)
+        for history_record in rm._make_history_records():
+            sm.send(history_record)
+            progress.subtask_pbar.update(task, advance=1)
+        progress.subtask_pbar.remove_task(task)
+
+
+def _handle_summary(sm: sender.SendManager, rm: RecordMaker, config: SendManagerConfig):
+    if config.summary:
+        sm.send(rm._make_summary_record())
+
+
+def _handle_terminal_output(
+    sm: sender.SendManager,
+    rm: RecordMaker,
+    config: SendManagerConfig,
+    run_identifier: str,
+):
+    if config.terminal_output:
+        task_name = f"Terminal Output {run_identifier}"
+        task = progress.subtask_pbar.add_task(task_name, total=None)
+        lines = rm.run.logs()
+        if lines is not None:
+            for line in lines:
+                sm.send(rm._make_output_record(line))
+                progress.subtask_pbar.update(task, advance=1)
+        progress.subtask_pbar.remove_task(task)
+
+
+def send_run_with_send_manager(
+    run: ImporterRun,
+    overrides: Optional[Dict[str, Any]] = None,
+    settings_override: Optional[Dict[str, Any]] = None,
+    config: Optional[SendManagerConfig] = None,
+) -> None:
+    if config is None:
+        config = SendManagerConfig()
+
+    # does this need to be here for pmap?
+    if overrides:
+        for k, v in overrides.items():
+            # `lambda: v` won't work!
+            # https://stackoverflow.com/questions/10802002/why-deepcopy-doesnt-create-new-references-to-lambda-function
+            setattr(run, k, lambda v=v: v)
+    rm = RecordMaker(run)
+
+    root_dir = rm.run_dir
+    settings = _make_settings(root_dir, settings_override)
+
+    record_q: queue.Queue = queue.Queue()
+    result_q: queue.Queue = queue.Queue()
+    interface = InterfaceQueue(record_q=record_q)
+    context_keeper = context.ContextKeeper()
+
+    run_identifier = f"{run.entity()}/{run.project()}/{run.run_id()}"
+
+    with sender.SendManager(
+        settings, record_q, result_q, interface, context_keeper
+    ) as sm:
+        _handle_run_record(sm, rm)
+        _handle_telem(sm, rm)
+        _handle_metadata(sm, rm, config)
+        _handle_use_artifacts(sm, rm, config, run_identifier)
```

Comment:
This means we will create the same artifact multiple times. This is likely fine since we deduplicate identical artifacts.